### ISSUE TITLE
[Bug] Evaluation on ConcatDataset

### ISSUE THREAD

**Issue by levan92** [open-mmlab/mmengine#730]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmengine/issues) and [Discussions](https://github.com/open-mmlab/mmengine/discussions) but cannot get the expected help.
- [X] The bug has not been fixed in the latest version(https://github.com/open-mmlab/mmengine).

### Environment

OrderedDict([('sys.platform', 'linux'), ('Python', '3.7.10 (default, Feb 26 2021, 18:47:35) [GCC 7.3.0]'), ('CUDA available', True), ('numpy_random_seed', 2147483648), ('GPU 0,1,2,3', 'NVIDIA A30'), ('CUDA_HOME', '/usr/local/cuda'), ('NVCC', 'Cuda compilation tools, release 11.1, V11.1.105'), ('GCC', 'gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0'), ('PyTorch', '1.9.0'), ('PyTorch compiling details', 'PyTorch built with:\n  - GCC 7.3\n  - C++ Version: 201402\n  - Intel(R) oneAPI Math Kernel Library Version 2021.2-Product Build 20210312 for Intel(R) 64 architecture applications\n  - Intel(R) MKL-DNN v2.1.2 (Git Hash 98be7e8afa711dc9b66c8ff3504129cb82013cdb)\n  - OpenMP 201511 (a.k.a. OpenMP 4.5)\n  - NNPACK is enabled\n  - CPU capability usage: AVX2\n  - CUDA Runtime 11.1\n  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37\n  - CuDNN 8.0.5\n  - Magma 2.5.2\n  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.1, CUDNN_VERSION=8.0.5, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.9.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, \n'), ('TorchVision', '0.10.0'), ('OpenCV', '4.6.0'), ('MMEngine', '0.1.0')])

### Reproduces the problem - code sample

```python
# in config file
data = dict(
    val=dict(
        type='ConcatDataset',
        datasets=[ 
            dict(
                type=dataset_type,
                ann_file=data_root + 'COCO2017/annotations/instances_val2017.json',
                img_prefix=data_root + 'COCO2017/val2017/',
                test_mode=True,
                classes=classes,
                pipeline=test_pipeline,
            ),
            dict(
                type=dataset_type,
                ann_file=data_root + 'custom/val.json',
                img_prefix=data_root + 'custom/images/',
                classes=classes,
                test_mode=True,
                pipeline=test_pipeline,
            ),
        ]
        ),
    )
```

### Reproduces the problem - command or script

Normal training command via `tools/dist_train.sh`

### Reproduces the problem - error message

See https://github.com/open-mmlab/mmdetection/issues/8785 
Basically, it seems like ConcatDataset has no support for evaluation (val/test) yet. Do let me know if this is already fixed or it's something that will be fixed very soon in your plans. Thank you!

**Comment by C1rN09** [comment]
Hi, I've read your issue in mmdetection. I suppose you are looking for a way to apply **the same metric** on **multiple datasets** and get  a **separate result on each dataset**, is it right?

`ConcatDataset` is designed for concatenation of multiple datasets, after which they cannot be distinguished from each other. Therefore, it may not suit your need.

I think you can implement a `ConcatDatasetWithKeys`, which will add a special key to the data from `__getitem__`. This will make your datasets distinguishable. However, if you do so, you may have to check that your model will parse the key correctly, and your metric should be reimplemented to be able to parse the key.

**Comment by levan92** [comment]
Thank you for you reply! 

> apply the same metric on multiple datasets and get a separate result on each dataset

Yes! Exactly that. In mmdet v2 / mmcv, `ConcatDataset` was able to do that. 

> I think you can implement a ConcatDatasetWithKeys, which will add a special key to the data from __getitem__

Got it! I will work along this line. Just wanted to make sure that there isn't a function/argument that already covered this in these suite of latest versions/library.

**Comment by C1rN09** [comment]
[An example in MMOCR](https://github.com/open-mmlab/mmocr/blob/1.x/mmocr/evaluation/evaluator/multi_datasets_evaluator.py) might be helpful to you.

If you find it not feasible, then I suppose you may be interested in rewriting a `loop` based on `ValLoop`, in which you can customize the logic to build multiple evaluators and match them with different datasets

**Comment by levan92** [comment]
> [An example in MMOCR](https://github.com/open-mmlab/mmocr/blob/1.x/mmocr/evaluation/evaluator/multi_datasets_evaluator.py) might be helpful to you.
> 
> If you find it not feasible, then I suppose you may be interested in rewriting a `loop` based on `ValLoop`, in which you can customize the logic to build multiple evaluators and match them with different datasets

Nice! I think the MMOCR example is highly relevant for my use case. Will take a look. Thank you for your guidance 😄

**Comment by wnakano** [comment]
@levan92 did you manage to create a module to perform ValLoop and TestLoop to work with ConcatDataset or multiples annotation files?

**Comment by gpadiolleau** [comment]
Any news about on evaluation with multiples annotation files ?
